plugins/outputs/cloudwatch/cloudwatch_test.go

```diff
@@ -33,30 +33,288 @@ func TestBuildDimensions(t *testing.T) {
+}
+
+// Test that the metric host tag is retained even when the global dimensions exceed the max number permitted
+func TestGlobalDimensionsMetricHostIsRetainedIfExceedMaxNumber(t *testing.T) {
+	assert := assert.New(t)
+
+	testPoint := testutil.TestMetric(1)
+
+	for key := range testPoint.Tags() {
+		testPoint.RemoveTag(key)
+	}
+
+	expectedHost := "some-host"
+	testPoint.AddTag("host", expectedHost)
+	testPoint.AddTag("shouldBeDropped", "OK")
+
+	globalDimensions := map[string]string{}
+
+	for i := 0; i <= MaxDimensions; i++ {
+		globalDimensions[fmt.Sprint(i)] = "Value"
+	}
+
+	dimensions := BuildDimensions(testPoint.Tags(), globalDimensions)
+
+	assert.Equal(MaxDimensions, len(dimensions), "Number of dimensions retained should be MaxDimensions")
+
+	var actualHost *string
+
+	for _, dimension := range dimensions {
+		assert.NotEqual("shouldBeDropped", *dimension.Value, "Global dimensions should be retained in preference to metric tags")
+
+		if *dimension.Name == "host" {
+			actualHost = dimension.Value
+			assert.Equal(expectedHost, *dimension.Value)
+		} else {
+			assert.Contains(globalDimensions, *dimension.Name, "Global dimensions should be retained")
+		}
+	}
+
+	assert.NotNil(actualHost, "Host tag should have been retained")
+}
+
+// Test that global dimensions override metric tags correctly
+func TestGlobalDimensionsOverride(t *testing.T) {
+	assert := assert.New(t)
+
+	testPoint := testutil.TestMetric(1)
+	testPoint.AddTag("Environment", "live")
+
+	globalDimensions := map[string]string{
+		"Environment": "test",
+		"Deployment":  "green",
+	}
+	dimensions := BuildDimensions(testPoint.Tags(), globalDimensions)
+
+	tagKeys := make([]string, len(testPoint.Tags())-1)
+	i := 0
+	for k := range testPoint.Tags() {
+		if k == "Environment" {
+			continue
+		}
 		tagKeys[i] = k
 		i += 1
 	}
 
 	sort.Strings(tagKeys)
 
-	if len(testPoint.Tags()) >= MaxDimensions {
-		assert.Equal(MaxDimensions, len(dimensions), "Number of dimensions should be less than MaxDimensions")
-	} else {
-		assert.Equal(len(testPoint.Tags()), len(dimensions), "Number of dimensions should be equal to number of tags")
+	assert.Equal(len(testPoint.Tags())+len(globalDimensions)-1, len(dimensions), "Conflicting global and metric dimensions should be merged")
+
+	for key, value := range globalDimensions {
+		var matchingDimension *cloudwatch.Dimension
+
+		for _, dimension := range dimensions {
+			if *dimension.Name == key {
+				matchingDimension = dimension
+				break
+			}
+		}
+
+		assert.NotNil(matchingDimension, "Global dimension should exist")
+		assert.Equal(value, *matchingDimension.Value, "Global dimension value should be equal")
 	}
 
 	for i, key := range tagKeys {
-		if i >= 10 {
-			break
+		assert.Equal(key, *dimensions[i+len(globalDimensions)].Name, "Key should be equal")
+		assert.Equal(testPoint.Tags()[key], *dimensions[i+len(globalDimensions)].Value, "Value should be equal")
+	}
+}
+
+// Test that global dimensions don't override the host metric tag
```

Comment:
Thanks will take a look today

Reply:
Was making a comment above about what you're referring to I assume - does it make any sense for someone to add `host` to their global dimensions? Is this something that we should guard against at translation time? I personally think it doesn't make sense to override existing behavior that we have for the host tag by letting customers drop a host tag in here. There is a check above in building the dimensions map to check if a global dimension key == host.. and as I'm thinking about it, I don't think that needs to be there if we guard against it during translation. Thoughts?

Reply:
The more I think about it, the more I think that excluding the `host` dimension from the global dimensions is a good idea. If someone has a specific use-case for overriding the host in the global dimensions then we can think again but otherwise it just seems like a source of unexpected behaviour / increased complexity / bugs etc.

I'll have a go at adding some checks for that in the translation layer — do you think adding something to exclude it in the schema is a good idea or would it be better to accept the invalid config and filter it out in the translation layer (possibly logging a warning that it's being dropped)?

Reply:
OK, I've just pushed a change that:

- Drops (silently) any `host` dimension from `global_dimensions` in the translator
- Drops (silently) any `host` dimension from `GlobalDimensions` in the CloudWatch output plugin

Strictly I didn't need to do the second but it seems moderately sensible to take a "belt and braces" approach as the additional logic required is very small (a single `key != "host"` check). Let me know what you think @SaxyPandaBear .

Reply:
Thanks. Sorry was busy yesterday will take a look today

Reply:
😓 was somehow even busier yesterday. I promise I will look by EOD Friday but not sure when I'll have time to sit down and review